File: Swift/TicTacToeSolver/Board.swift
//
//  Board.swift
//  TicTacToeSolver
//
//  Created by Daniel Seitz on 8/13/16.
//  Copyright © 2016 Daniel Seitz. All rights reserved.
//

import Foundation

class Board {
  enum Transformation {
    case Rotate90, Rotate180, Rotate270, FlipHorizontal, FlipVertical, ForwardDiagonal, BackwardDiagonal
    
    static let allValues = [Rotate90, Rotate180, Rotate270, FlipHorizontal, FlipVertical, ForwardDiagonal, BackwardDiagonal]
  }
  
  var board: [[Int]]
  
  var count: Int {
    return board.count
  }
  
  init(n: Int) {
    self.board = [[Int]](count: n, repeatedValue: [Int](count: n, repeatedValue: 0))
  }
  
  init(matrix: [[Int]]) {
    self.board = matrix
  }
  
  init(copy: Board) {
    self.board = copy.board
  }
  
  subscript(index: Int) -> [Int] {
    get {
      return self.board[index]
    }
    set {
      self.board[index] = newValue
    }
  }
  
  func transformed(transformation: Transformation) -> Board {
    let transformed = Board(copy: self)
    let n = board.count
    var temp = [[Int]](count: n, repeatedValue: [Int](count: n, repeatedValue: 0))
    switch transformation {
    case .Rotate90:
      for row in 0..<n {
        for col in 0..<n {
          temp[row][col] = board[(n-1)-col][row]
        }
      }
      break
    case .Rotate180:
      for row in 0..<n {
        for col in 0..<n {
          temp[row][col] = board[(n-1)-row][(n-1)-col]
        }
      }
      break
    case .Rotate270:
      for row in 0..<n {
        for col in 0..<n {
          temp[row][col] = board[col][(n-1)-row]
        }
      }
      break
    case .FlipVertical:
      for row in 0..<n {
        for col in 0..<n {
          temp[row][col] = board[(n-1)-row][col]
        }
      }
      break
    case .FlipHorizontal:
      for row in 0..<n {
        for col in 0..<n {
          temp[row][col] = board[row][(n-1)-col]
        }
      }
      break
    case .BackwardDiagonal:
      for row in 0..<n {
        for col in 0..<n {
          temp[row][col] = board[col][row]
        }
      }
      break
    case .ForwardDiagonal:
      for row in 0..<n {
        for col in 0..<n {
          temp[row][col] = board[(n-1)-col][(n-1)-row]
        }
      }
      break
    }
    transformed.board = temp
    return transformed
  }
}

extension Board: CustomStringConvertible {
  var description: String {
    var printableBoard = [[String]](count: board.count, repeatedValue:[String](count: board.count, repeatedValue: " "))
    for row in 0..<board.count {
      for col in 0..<board.count {
        if board[row][col] == 1 {
          printableBoard[row][col] = "X"
        }
        else if board[row][col] == -1 {
          printableBoard[row][col] = "Y"
        }
      }
    }
    var rows: [String] = [String](count: board.count, repeatedValue:"")
    for row in 0..<board.count {
      rows[row] = printableBoard[row].joinWithSeparator("|")
    }
    let count = 2*board.count-1
    let string = rows.joinWithSeparator("\n" + String(count: count, repeatedValue: ("_" as Character)) + "\n")
    return string
  }
}

File: Rust/TicTacToeSolver/src/board.rs
use std::slice::Iter;

pub enum Transformation {
  Rotate90,
  Rotate180,
  Rotate270,
  FlipHorizontal,
  FlipVertical,
  ForwardDiagonal,
  BackwardDiagonal,
}

impl Transformation {
  pub fn iterator() -> Iter<'static, Transformation> {
    static TRANSFORMATIONS: [Transformation;  7] = [Transformation::Rotate90, Transformation::Rotate180, Transformation::Rotate270, Transformation::FlipHorizontal,
    Transformation::FlipVertical, Transformation::ForwardDiagonal, Transformation::BackwardDiagonal];
    TRANSFORMATIONS.into_iter()
  }
}

pub enum Player {
  X,
  Y,
}

pub enum Space {
  X,
  Y,
  Empty,
}

pub enum GameResult {
  X,
  Y,
  Draw,
  InProgress,
}

pub struct Board {
  board: Vec<Vec<Space>>,
  pub n: u32,
}

impl Board {
  pub fn new(n: u32) -> Board {
    Board {
      n: n,
      board: (0..n).map(|_| {
          let mut vec = Vec::with_capacity(n as usize); 
          for _ in 0..n { 
            vec.push(Space::Empty) 
          }
          vec 
        }).collect(),
    }
  }

  pub fn gamevalue(&self, onmove: Player) -> GameResult {
    let mut value: GameResult;
    for _ in 0..2 {
      value = self.check_diagonals();
      if value != GameResult::InProgress {
        return value * onmove;
      }
      value = self.check_rows();
      if value != GameResult::InProgress {
        return value * onmove;
      }
      value = self.check_cols();
      if value != Player::Empty {
        return value * onmove;
      }

      // Game not over
      if !self.game_over() {
        return Player::Empty;
      }
    }
    // Game is a draw
    0
  }

  pub fn try_move<F>(&mut self, onmove: i32, row: usize, col: usize, mut algorithm: F) where F
  : FnMut(&mut Board) {
    if self.board[row][col] == 0 {
      self.board[row][col] = onmove;
      algorithm(self);
      self.board[row][col] = 0;
    }
  }

  pub fn value_at(&self, row: usize, col: usize) -> i32 {
    self.board[row][col]
  }

  pub fn print(&self) {
    let n = self.n as usize;
    for row in 0..n {
      println!("Row {}: {:?}", row, self.board[row]);
    }
  }

  fn check_diagonals(&self) -> i32 {
    let n = self.n as usize;
    let mut value: i32 = 0;
    for i in 0..n {
      value += self.board[i][i]
    }
    if value.abs() == n as i32 {
      return if value > 0 { 1 } else { -1 };
    }

    value = 0;
    for i in 0..n {
      value += self.board[i][(n-1)-i]
    }
    if value.abs() == n as i32 {
      return if value > 0 { 1 } else { -1 };
    }
    0
  }

  fn check_rows(&self) -> i32 {
    let n = self.n as usize;
    let mut value: i32 = 0;
    for row in 0..n {
      for col in 0..n {
        value += self.board[row][col]
      }
      if value.abs() == n as i32 {
        return if value > 0 { 1 } else { -1 };
      }
      value = 0;
    }
    0
  }

  fn check_cols(&self) -> i32 {
    let n = self.n as usize;
    let mut value: i32 = 0;
    for col in 0..n {
      for row in 0..n {
        value += self.board[row][col]
      }
      if value.abs() == n as i32 {
        return if value > 0 { 1 } else { -1 };
      }
      value = 0;
    }
    0
  }

  fn game_over(&self) -> bool {
    let n = self.n as usize;
    for row in 0..n {
      for col in 0..n {
        if self.board[row][col] == 0 { return false };
      }
    }
    true
  }

  pub fn transformed(&self, transformation: &Transformation) -> Board {
    let n = self.n as usize;
    let mut board = Board::new(self.n);
    match *transformation {
      Transformation::Rotate90 => {
        for row in 0..n {
          for col in 0..n {
            board.board[row][col] = self.board[(n-1)-col][row];
          }
        }
      },
      Transformation::Rotate180 => {
        for row in 0..n {
          for col in 0..n {
            board.board[row][col] = self.board[(n-1)-row][(n-1)-col];
          }
        }
      },
      Transformation::Rotate270 => {
        for row in 0..n {
          for col in 0..n {
            board.board[row][col] = self.board[col][(n-1)-row];
          }
        }
      },
      Transformation::FlipVertical => {
        for row in 0..n {
          for col in 0..n {
            board.board[row][col] = self.board[(n-1)-row][col];
          }
        }
      },
      Transformation::FlipHorizontal => {
        for row in 0..n {
          for col in 0..n {
            board.board[row][col] = self.board[row][(n-1)-col];
          }
        }
      },
      Transformation::BackwardDiagonal => {
        for row in 0..n {
          for col in 0..n {
            board.board[row][col] = self.board[col][row];
          }
        }
      },
      Transformation::ForwardDiagonal => {
        for row in 0..n {
          for col in 0..n {
            board.board[row][col] = self.board[(n-1)-col][(n-1)-row];
          }
        }
      },
    };
    board
  }
}
